### PR TITLE
[image][Android] Add automatic asset downscaling to improve performance

### LINKED ISSUE #21491: [expo-image][Android] performance UI fps drop with some specific jpeg images
### Minimal reproducible example

https://github.com/zhelezkov/expo-image-performance-issue

### Summary

Hey all,
Thanks for releasing expo-image, it's fantastic library and I hope to use it in prod. 

Unfortunately I have faced a major performance issue with `expo-image` and specific combination of jpeg images. I know it might sounds crazy, and I didn't believe my eyes when I found it, **BUT** some specific combination of jpeg images causes a huge performance drop during scroll, from 120 FPS to ~40 FPS.

In repro, I left comment about images that causes a performance drop https://github.com/zhelezkov/expo-image-performance-issue/blob/main/App.js#L20-L21. If you try to comment it out, you'll see that performance is fine. 

I almost certain that the problem is in the `expo-image`, because if I switch to `Image` component from `react-native`, it works pretty well and I don't see such major performance drops. (https://github.com/zhelezkov/expo-image-performance-issue/blob/main/App.js#L30)
Ideally I would expect `expo-image` to work faster or at least with the same performance as default `Image` component from `react-native`.

I have also checked same images with `react-native-fast-image` and it looks fine, no performance drops

<details>
<summary>Please also find video with performance problem. FPS drops from 120 to 40.</summary>

https://user-images.githubusercontent.com/5163918/222594858-3bcb6dfd-6e23-4f8f-a6ad-11d70d81d8a1.MOV
</details>

<details>
<summary>
And same project, but with default Image component from react-native. no performance drop
</summary>

https://user-images.githubusercontent.com/5163918/222594838-5c03e32e-400c-4981-ac55-918e005edb12.MOV
</details>

I have attached minimal repro, you can run it with expo go. 
This problem doesn't reproduce on iOS.

Please lmk if you need any other information

### Environment
```
  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 13.2.1
      Shell: 5.8.1 - /bin/zsh
    Binaries:
      Node: 18.12.1 - ~/.nvm/versions/node/v18.12.1/bin/node
      Yarn: 1.22.19 - ~/.nvm/versions/node/v18.12.1/bin/yarn
      npm: 8.19.2 - ~/.nvm/versions/node/v18.12.1/bin/npm
      Watchman: 2023.02.13.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.11.3 - /Users/nz/.rbenv/shims/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 22.2, iOS 16.2, macOS 13.1, tvOS 16.1, watchOS 9.1
    IDEs:
      Android Studio: Electric Eel 2022.1.1 Patch 1 Electric Eel 2022.1.1 Patch 1
      Xcode: 14.2/14C18 - /usr/bin/xcodebuild
    npmPackages:
      expo: ~48.0.4 => 48.0.4
      react: 18.2.0 => 18.2.0
      react-native: 0.71.3 => 0.71.3
    npmGlobalPackages:
      eas-cli: 3.7.2
    Expo Workflow: managed
```

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

Closes https://github.com/expo/expo/issues/21491.

# How

An automatic asset downscaling feature has been added to considerably decrease the memory and CPU usage of large images. The chosen downsampling technique has been set as the default option to align with iOS behavior and ensure no negative impact on users. Nonetheless, there is an option to disable it if desired.

# Test Plan

- bare-expo ✅
- https://github.com/zhelezkov/expo-image-performance-issue ✅

### CODE REVIEW

**Comment by fobos531** [comment]
Is this PR aimed to (among others) solving issues of this nature? Just started seeing them in Sentry.

CleanShot 2023-03-09 at 14 45 34@2x

**Comment by lukmccall** [comment]
> Is this PR aimed at (among others) solving issues of this nature? Just started seeing them in Sentry.

It wasn't the main reason behind it, but it may help in your case too.

File: packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageModule.kt [hunk 1]
```diff
@@ -162,6 +162,10 @@ class ExpoImageModule : Module() {
         view.recyclingKey = recyclingKey
       }
 
+      Prop("enableDownscaling") { view: ExpoImageViewWrapper, enableDownscaling: Boolean? ->
```
Comment (tsapeta) [inline packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageModule.kt:165]: What about `allowDownscaling` or `disableDownscaling`?
For me "enable" means that downscaling will always be used, but in fact it's done only when necessary.

File: packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageViewWrapper.kt [hunk 2]
```diff
@@ -468,6 +482,71 @@ class ExpoImageViewWrapper(context: Context, appContext: AppContext) : ExpoView(
       }
       newTarget.hasSource = sourceToLoad != null
 
+      val downsampleStrategy = if (enableDownscaling) {
+        object : DownsampleStrategy() {
+          var wasTrigger = false
```
Comment (tsapeta) [inline packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageViewWrapper.kt:487]:
```suggestion
          var wasTriggered = false
```

File: packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageViewWrapper.kt [hunk 3]
```diff
@@ -468,6 +482,71 @@ class ExpoImageViewWrapper(context: Context, appContext: AppContext) : ExpoView(
...
+
+            // The size of the container is unknown, we don't know what to do, so we just run the default scale.
+            if (requestedWidth == SIZE_ORIGINAL || requestedHeight == SIZE_ORIGINAL) {
+              return 1f
+            }
+
+            val rW = requestedWidth.toFloat()
+            val rH = requestedHeight.toFloat()
+            val sW = sourceWidth.toFloat()
+            val sH = sourceHeight.toFloat()
```
Comment (tsapeta) [inline packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageViewWrapper.kt:511]: Using such short names for variables is a bit odd 😅 Can't we just use these `.toFloat()` in the `when` block?

File: packages/expo-image/src/Image.types.ts [hunk 4]
```diff
@@ -255,6 +255,17 @@ export interface ImageProps extends ViewProps {
    * @platform ios 16.0+
    */
   enableLiveTextInteraction?: boolean;
+
+  /**
+   * Whether the image should be downscaled to match the size of the view container.
+   * Turning off this functionality could negatively impact the application's performance, particularly when working with large assets.
+   * However, it would result in smoother image resizing, and end-users would always have access to the highest possible asset quality.
+   *
+   * Downscaling is disable when the `contentFit` is set to `none` or `fill`.
```
Comment (tsapeta) [inline packages/expo-image/src/Image.types.ts:264]:
```suggestion
   * Downscaling is never used when the `contentFit` prop is set to `none` or `fill`.
```

**Review by tsapeta — APPROVED** [review]

**Pull request merged**